app/src/main/java/eu/theinvaded/mastondroid/ui/activity/LoginActivity.java

```diff
@@ -70,49 +94,67 @@ public void startMainActivity() {
                 .putString(getString(R.string.CURRENT_USERNAME), account.username)
                 .apply();
     }
 
-
-
     @Override
-    public void showLoginError() {
-        Toast.makeText(context, R.string.login_failed_message, Toast.LENGTH_SHORT).show();
+    public void signIn(String node) {
+        Map<String, String> parameters = new HashMap<>();
+        parameters.put("client_id", preferences.getString(getString(R.string.clientId), ""));
+        parameters.put("redirect_uri", "eu.theinvaded.mastondroid://oauth2redirect/");
+        parameters.put("response_type", "code");
+        parameters.put("scope", "read write follow");
+        String queryParameters;
+        try {
+            queryParameters = toQueryString(parameters);
+        } catch (UnsupportedEncodingException e) {
+            //TODO: No clue how to handle this error case??
+            assert(false);
+            return;
+        }
+        String url = "https://mastodon.social/oauth/authorize?" + queryParameters;
+        Intent viewIntent = new Intent("android.intent.action.VIEW", Uri.parse(url));
+        startActivity(viewIntent);
     }
 
     @Override
-    public void setNoUsernameError() {
-        binding.usernameLayout.setErrorEnabled(true);
-        binding.usernameLayout.setError(getString(R.string.NO_USERNAME_ERROR));
+    public boolean checkAppRegistered() {
+        return preferences.contains(getString(R.string.clientId))
+                && preferences.contains(getString(R.string.clientSecret));
     }
 
     @Override
-    public void setNoPaswordError() {
-        binding.passwordLayout.setErrorEnabled(true);
-        binding.passwordLayout.setError(getString(R.string.NO_PASSWORD_ERROR));
+    public void registerApp(String clientId, String clientSecret) {
+        preferences.edit()
+                .putString(getString(R.string.clientId), clientId)
+                .putString(getString(R.string.clientSecret), clientSecret)
+                .apply();
     }
 
     @Override
-    public void clearNoUsernameError() {
-        binding.usernameLayout.setErrorEnabled(false);
+    public void authorizeApp(String accessToken) {
+        preferences.edit()
+                .putString(getString(R.string.authKey), accessToken)
+                .apply();
+
+        startMainActivity();
     }
 
-    @Override
-    public void clearNoPasswordError() {
-        binding.passwordLayout.setErrorEnabled(false);
+    private String toQueryString(Map<String, String> parameters)
+            throws UnsupportedEncodingException {
+        StringBuilder s = new StringBuilder();
+        String between = "";
+        for (Map.Entry<String, String> entry : parameters.entrySet()) {
+            s.append(between);
+            s.append(URLEncoder.encode(entry.getKey(), "UTF-8"));
+            s.append("=");
+            s.append(URLEncoder.encode(entry.getValue(), "UTF-8"));
+            between = "&";
+        }
+        return s.toString();
```

Comment:
Probably could be improved with String.join.

Shouldn't this use some kind of common URL constructing library though?